src/test/validation_block_tests.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <iomanip>
+#include <memory>
+#include <sstream>
+
+#include <boost/test/unit_test.hpp>
+
+#include <chainparams.h>
+#include <consensus/merkle.h>
+#include <consensus/validation.h>
+#include <miner.h>
+#include <random.h>
+#include <test/test_bitcoin.h>
+#include <validation.h>
+#include <validationinterface.h>
+
+struct RegtestingSetup : public TestingSetup {
+    RegtestingSetup() : TestingSetup(CBaseChainParams::REGTEST) {}
+};
+
+BOOST_FIXTURE_TEST_SUITE(validation_block_tests, RegtestingSetup)
+
+struct TestSubscriber : public CValidationInterface {
+    uint256 m_expected_tip;
+
+    TestSubscriber(uint256 tip) : m_expected_tip(tip) {}
+
+    void UpdatedBlockTip(const CBlockIndex* pindexNew, const CBlockIndex* pindexFork, bool fInitialDownload)
+    {
+        BOOST_CHECK_EQUAL(m_expected_tip, pindexNew->GetBlockHash());
+    }
+
+    void BlockConnected(const std::shared_ptr<const CBlock>& block, const CBlockIndex* pindex, const std::vector<CTransactionRef>& txnConflicted)
+    {
+        BOOST_CHECK_EQUAL(m_expected_tip, block->hashPrevBlock);
+        BOOST_CHECK_EQUAL(m_expected_tip, pindex->pprev->GetBlockHash());
+
+        m_expected_tip = block->GetHash();
+    }
+
+    void BlockDisconnected(const std::shared_ptr<const CBlock>& block)
+    {
+        BOOST_CHECK_EQUAL(m_expected_tip, block->GetHash());
+
+        m_expected_tip = block->hashPrevBlock;
+    }
+};
+
+std::shared_ptr<CBlock> Block(const uint256& prev_hash)
+{
+    static int i = 0;
+    static uint64_t time = Params().GenesisBlock().nTime;
+
+    CScript pubKey;
+    pubKey << i++ << OP_TRUE;
+
+    auto ptemplate = BlockAssembler(Params()).CreateNewBlock(pubKey, false);
+    CBlock* pblock = new CBlock(ptemplate->block);
+    pblock->hashPrevBlock = prev_hash;
+    pblock->nTime = ++time;
+
+    return std::make_shared<CBlock>(*pblock);
+}
+
+std::shared_ptr<CBlock> FinalizeBlock(std::shared_ptr<CBlock> pblock)
+{
+    pblock->hashMerkleRoot = BlockMerkleRoot(*pblock);
+
+    while (!CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())) {
+        ++(pblock->nNonce);
+    }
+
+    return pblock;
+}
+
+// construct a valid block
+const std::shared_ptr<const CBlock> GoodBlock(const uint256& prev_hash)
+{
+    return FinalizeBlock(Block(prev_hash));
+}
+
+// construct an invalid block (but with a valid header)
+const std::shared_ptr<const CBlock> BadBlock(const uint256& prev_hash)
+{
+    auto pblock = Block(prev_hash);
+
+    // a second coinbase will make this block invalid
+    pblock->vtx.push_back(pblock->vtx[0]);
```

Comment:
This method of constructing a bad block is not ideal, because this kind of invalidity is detected well before ConnectBlock.  In order to get better code coverage, I think it would be better to generate different kinds of invalid blocks, to test failure at different points in the validation process.

(As it is,  I believe this method of making an invalid block is detectable in CheckBlock (merkle-tree fails validation due to duplicate transactions).  It would also fail in CheckBlock for having more than one coinbase, and it would fail in ContextualCheckBlock for having an invalid witness commitment.)